src/audio/google_rtc_audio_processing.c

```diff
@@ -371,50 +371,59 @@ static int google_rtc_audio_processing_copy(struct comp_dev *dev)
 
 	buffer_stream_invalidate(cd->aec_reference, num_aec_reference_bytes);
 
-	aec_reference_buff_frag = 0;
-	for (frame = 0; frame < num_aec_reference_frames; frame++) {
-		for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
-			src = audio_stream_read_frag_s16(&cd->aec_reference->stream,
-							 aec_reference_buff_frag + channel);
-			cd->aec_reference_buffer[cd->num_aec_reference_channels *
-				cd->aec_reference_frame_index + channel] = *src;
+	remain = num_aec_reference_frames * cd->num_aec_reference_channels;
+	while (remain) {
+		nmax = audio_stream_samples_without_wrap_s16(&cd->aec_reference->stream, ref);
+		n  = MIN(remain, nmax);
+		for (i = 0; i < n; i += cd->num_aec_reference_channels) {
+			j = cd->num_aec_reference_channels * cd->aec_reference_frame_index;
+			for (channel = 0; channel < cd->num_aec_reference_channels; ++channel) {
+				cd->aec_reference_buffer[j++] = *ref;
+				ref++;
+			}
+			++cd->aec_reference_frame_index;
+
+			if (cd->aec_reference_frame_index == cd->num_frames) {
+				GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
+									    cd->aec_reference_buffer);
+				cd->aec_reference_frame_index = 0;
+			}
 		}
-		++cd->aec_reference_frame_index;
-
-		if (cd->aec_reference_frame_index == cd->num_frames) {
-			GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state,
-								    cd->aec_reference_buffer);
-			cd->aec_reference_frame_index = 0;
-		}
-		aec_reference_buff_frag += cd->aec_reference->stream.channels;
+		remain -= n;
+		ref = audio_stream_wrap(&cd->aec_reference->stream, ref);
 	}
 	comp_update_buffer_consume(cd->aec_reference, num_aec_reference_bytes);
 
 	comp_get_copy_limits_with_lock(cd->raw_microphone, cd->output, &cl);
 	buffer_stream_invalidate(cd->raw_microphone, cl.source_bytes);
 
-	raw_microphone_buff_frag = 0;
-	output_buff_frag = 0;
-	for (frame = 0; frame < cl.frames; frame++) {
-		src = audio_stream_read_frag_s16(&cd->raw_microphone->stream,
-										 raw_microphone_buff_frag);
-		cd->raw_mic_buffer[cd->raw_mic_buffer_index] = *src;
-		++cd->raw_mic_buffer_index;
-
-		dst = audio_stream_read_frag_s16(&cd->output->stream, output_buff_frag);
-		*dst = cd->output_buffer[cd->output_buffer_index];
-		++cd->output_buffer_index;
-
-		if (cd->raw_mic_buffer_index == cd->num_frames) {
-			GoogleRtcAudioProcessingProcessCapture_int16(
-				cd->state, cd->raw_mic_buffer,
-				cd->output_buffer);
-			cd->output_buffer_index = 0;
-			cd->raw_mic_buffer_index = 0;
+	remain = cl.frames;
```

Comment:
we could also write: `remain = cl.frames * cd->num_capture_channels` where we initialize `cd->num_capture_channels = 1` in `google_rtc_audio_processing_create`
I think that would make more sense.

Technically we could have (That might be for a follow up PR) different number of channels for capture input and capture output.

Reply:
I leave that up to you. I'm OK to change but counting frames here preserves possibility for different number of source and sink channes. What is your preference?

Reply:
I would prefer counting frame everywhere converting to samples when needed.

Reply:
OK, here the frames counting is needed. The implementation is not most efficient, there's divide and a switch case in audio_stream_frame_bytes().
```
static inline uint32_t
audio_stream_frames_without_wrap(const struct audio_stream *source,
				 const void *ptr)
{
	uint32_t bytes = audio_stream_bytes_without_wrap(source, ptr);
	uint32_t frame_bytes = audio_stream_frame_bytes(source);

	return bytes / frame_bytes;
}

```
While this function is a lot lighter, so I used in reference where it was possible.

```
static inline int
audio_stream_samples_without_wrap_s16(const struct audio_stream *source, const void *ptr)
{
	int to_end = (int16_t *)source->end_addr - (int16_t *)ptr;

	assert((intptr_t)source->end_addr >= (intptr_t)ptr);
	return to_end;
}

```